pkg/trtl/trtl.go

```diff
@@ -1,156 +1,566 @@
+		log.Error().Err(err).Str("key", string(in.Key)).Msg("unable to delete object")
+		return nil, status.Error(codes.Internal, err.Error())
 	}
-	peers.RegisterPeerManagementServer(s.srv, s.peers)
 
-	// TODO: initialize the Replica service
+	out = &pb.DeleteReply{Success: true}
+
+	if in.Options != nil && in.Options.ReturnMeta {
+		out.Meta = returnMeta(object)
+	}
 
-	return s, nil
+	return out, nil
 }
 
-// Serve gRPC requests on the specified bind address.
-func (t *Server) Serve() (err error) {
-	// Catch OS signals for graceful shutdowns
-	quit := make(chan os.Signal, 1)
-	signal.Notify(quit, os.Interrupt)
-	go func() {
-		<-quit
-		t.echan <- t.Shutdown()
-	}()
-
-	// Run management routines only if we're not in maintenance mode
-	if t.conf.Maintenance {
-		log.Warn().Msg("starting trtl in maintenance mode")
+// Iter is a unary request to fetch a materialized collection of key/value pairs based
+// on a shared prefix. If no prefix is specified an entire namespace may be returned.
+// This RPC supports pagination to ensure that replies do not get too large. The default
+// page size is 100 items, though this can be modified in the options. The next page
+// token in the result will contain the next page to request, or will be empty if there
+// are no more results to be supplied.
+//
+// Note that there are no snapshot guarantees during iteration, meaning that if the
+// underlying database changes between requests, these changes could be reflected during
+// iteration. For snapshot isolation in iteration, use the Cursor RPC.
+//
+// There are several options that modulate the Iter response:
+//   - return_meta: each key/value pair will contain the object metadata
+//   - iter_no_keys: each key/value pair will not have a key associated with it
+//   - iter_no_values: each key/value pair will not have a value associated with it
+//   - page_token: the page of results that the user wishes to fetch
+//   - page_size: the number of results to be returned in the request
+func (h *TrtlService) Iter(ctx context.Context, in *pb.IterRequest) (out *pb.IterReply, err error) {
+	// Ensure the namespace is not reserved
+	if _, found := reservedNamespaces[in.Namespace]; found {
+		log.Warn().Str("namespace", in.Namespace).Msg("cannot use reserved namespace")
+		return nil, status.Error(codes.PermissionDenied, "cannot used reserved namespace")
+	}
+
+	// Load the options from the request
+	var opts *pb.Options
+	if in.Options != nil {
+		opts = in.Options
 	} else {
-		// These services should not run in maintenance mode
-		// Run the Gossip background routine
-		go t.replica.AntiEntropy()
+		// Create empty options
+		opts = &pb.Options{}
 	}
 
-	// Listen for TCP requests
-	var sock net.Listener
-	if sock, err = net.Listen("tcp", t.conf.BindAddr); err != nil {
-		log.Error().Err(err).Str("bindaddr", t.conf.BindAddr).Msg("could not listen on given bindaddr")
-		return err
+	// Load defaults into options
+	if opts.PageSize == 0 {
+		opts.PageSize = 100
```

Comment:
I wonder if we should pull this out and make it a global variable at the top of the file so that it's easier to find in the future if we need to do any troubleshooting around pagination?

Reply:
Yes, you're right - I've done that to make it more obvious. 

However, I think that when we finally hone in on what we think will be the stable API for options, we should create a `LoadOptions` method or similar to load options with the defaults so all the defaults across all methods are in one place. Probably it should also do the work of merging the options specified by the user as well to replace all of lines 193-205 and the duplicated code in the other RPCs :-)  -- I don't want to do this though until we're feeling confident that `Options` is how we're going to move forward and not flattening these arguments into the request RPCs.